File: backend/src/main/java/com/mvp/cybergi/services/ModuleService.java
package com.mvp.cybergi.services;

import com.mvp.cybergi.model.entity.Lesson;
import com.mvp.cybergi.model.entity.Module;
import com.mvp.cybergi.repository.ModuleRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class ModuleService {

    @Autowired
    private ModuleRepository moduleRepository;

    // Create a new module
    public Module createModule(Module module) {
        return moduleRepository.save(module);
    }

    // Retrieve all modules
    public List<Module> getAllModules() {
        return moduleRepository.findAll();
    }

    // Retrieve featured modules
    public List<Module> getFeaturedModules() {
        return moduleRepository.findByFeatured(true); // Return only featured modules
    }

    // Retrieve a module by its ID
    public Optional<Module> getModuleById(Long id) {
        return moduleRepository.findById(id);
    }

    // Update an existing module
    public Module updateModule(Long id, Module updatedModule) {
        return moduleRepository.findById(id)
                .map(module -> {
                    module.setTitle(updatedModule.getTitle());
                    module.setDescription(updatedModule.getDescription());

                    // Clear the existing lessons and add the updated lessons
                    module.getLessons().clear();
                    module.getLessons().addAll(updatedModule.getLessons());

                    // Set the module reference for each lesson to ensure consistency
                    for (Lesson lesson : module.getLessons()) {
                        lesson.setModule(module);
                    }

                    return moduleRepository.save(module);
                }).orElseThrow(() -> new RuntimeException("Module not found"));
    }

    // Delete a module by its ID
    public void deleteModule(Long id) {
        moduleRepository.deleteById(id);
    }

    // Delete all modules
    public void deleteAllModules() {
        moduleRepository.deleteAll();
    }
}

File: frontend/src/app/services/module.service.ts
// src/app/services/module.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class ModuleService {
  private apiUrl = environment.apiUrl + '/modules';

  constructor(private http: HttpClient) {}

  createModule(module: any): Observable<any> {
    return this.http.post(`${this.apiUrl}/create`, module);
  }

  getAllModules(): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/all`);
  }

  getModuleById(moduleId: number): Observable<any> {
    return this.http.get(`${this.apiUrl}/${moduleId}`);
  }

  updateModule(moduleId: number, module: any): Observable<any> {
    return this.http.put(`${this.apiUrl}/update/${moduleId}`, module);
  }

  deleteModule(moduleId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/delete/${moduleId}`);
  }

  getFeaturedModules(): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/featured`);
  }
}
